src/core/utils/link_metrics_manager.cpp

```diff
@@ -0,0 +1,396 @@
+    {
+        Start();
+    }
+    else
+    {
+        Stop();
+    }
+
+exit:
+    return;
+}
+
+Error LinkMetricsManager::GetLinkMetricsValueByExtAddr(const Mac::ExtAddress      &aExtAddress,
+                                                       LinkMetrics::MetricsValues &aMetricsValues)
+{
+    Error    error = kErrorNone;
+    Subject *subject;
+
+    subject = mSubjectList.FindMatching(aExtAddress);
+    VerifyOrExit(subject != nullptr, error = kErrorNotFound);
+
+    aMetricsValues.mLinkMarginValue = subject->mData.mLinkMargin;
+    aMetricsValues.mRssiValue       = subject->mData.mRssi;
+
+exit:
+    return error;
+}
+
+void LinkMetricsManager::Start(void)
+{
+    LinkMetrics::Initiator &initiator = Get<LinkMetrics::Initiator>();
+
+    VerifyOrExit(mEnabled && Get<Mle::Mle>().IsAttached());
+
+    initiator.SetMgmtResponseCallback(HandleMgmtResponse, this);
+    initiator.SetEnhAckProbingCallback(HandleEnhAckIe, this);
+
+    mTimer.Start(kTimeBeforeStartMilliSec);
+exit:
+    return;
+}
+
+void LinkMetricsManager::Stop(void)
+{
+    LinkMetrics::Initiator &initiator = Get<LinkMetrics::Initiator>();
+
+    mTimer.Stop();
+
+    initiator.SetMgmtResponseCallback(nullptr, nullptr);
+    initiator.SetEnhAckProbingCallback(nullptr, nullptr);
+
+    UnregisterAllSubjects();
+    ReleaseAllSubjects();
+}
+
+void LinkMetricsManager::Update(void)
+{
+    UpdateSubjects();
+    UpdateLinkMetricsStates();
+}
+
+// This method updates the Link Metrics Subject in the subject list. It adds new neighbors to the list.
+void LinkMetricsManager::UpdateSubjects(void)
+{
+    Neighbor::Info         neighborInfo;
+    otNeighborInfoIterator iterator = OT_NEIGHBOR_INFO_ITERATOR_INIT;
+
+    while (Get<NeighborTable>().GetNextNeighborInfo(iterator, neighborInfo) == kErrorNone)
+    {
+        // if not in the subject list, allocate and add
+        if (!mSubjectList.ContainsMatching(AsCoreType(&neighborInfo.mExtAddress)))
+        {
+            Subject *subject = mPool.Allocate();
+
+            VerifyOrExit(subject != nullptr);
+
+            subject->Clear();
+            subject->mExtAddress = AsCoreType(&neighborInfo.mExtAddress);
+            IgnoreError(mSubjectList.Add(*subject));
+        }
+    }
+
+exit:
+    return;
+}
+
+// This method updates the state and take corresponding actions for all subjects and removes stale subjects.
+void LinkMetricsManager::UpdateLinkMetricsStates(void)
+{
+    ot::LinkedList<Subject> staleSubjects;
+
+    Subject *entry;
+    Subject *prev;
+    Subject *next;
+
+    for (prev = nullptr, entry = mSubjectList.GetHead(); entry != nullptr; entry = next)
+    {
+        Error error = entry->UpdateState(GetInstance());
+
+        next = entry->GetNext();
+
+        if (error == kErrorUnknownNeighbor || error == kErrorNotCapable)
+        {
+            mSubjectList.PopAfter(prev);
+            IgnoreError(staleSubjects.Add(*entry));
+        }
+        else
+        {
+            prev = entry;
+        }
+    }
```

Comment:
The code here is correct but it is always tricky one to get right (iterating over list while removing some items from it).

We can consider using `LinkedList::RemoveAllMatching()` to do something similar (iterate over all check if the match certain condition, remove them from list and add them to new `staleList`?

We may need to define a `Subject::Matches()` with some dummy `Indicator` type (maybe an `enum` type or can use `Instance` so it can be use by `UpdateState()` call ?) that then does call `UpdateState()`  and check the `error` to convert to a `bool`.

Reply:
Good point!